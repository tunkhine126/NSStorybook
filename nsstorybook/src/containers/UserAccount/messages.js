/*
 * UserAccount Messages
 *
 * This contains all the text for the UserAccount container.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.containers.UserAccount';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the UserAccount container!',
  },
  accSettingsHeader: {
    id: `${scope}.accSettingsHeader`,
    defaultMessage: 'Account Settings',
  },
  accountUpdated: {
    id: `${scope}.accountUpdated`,
    defaultMessage: 'Account Settings Updated.',
  },
  confirmChanges: {
    id: `${scope}.confirmChanges`,
    defaultMessage: 'Confirm Changes',
  },
  confirmPassword: {
    id: `${scope}.confirmPassword`,
    defaultMessage: 'Confirm Your New Password',
  },
  currentPassword: {
    id: `${scope}.currentPassword`,
    defaultMessage: 'Current Password',
  },
  defaultSurveyLang: {
    id: `${scope}.defaultSurveyLang`,
    defaultMessage: 'Default Survey Language',
  },
  defaultLang: {
    id: `${scope}.defaultLang`,
    defaultMessage: 'Default Language',
  },
  emailChange: {
    id: `${scope}.emailChange`,
    defaultMessage: 'To change your email please contact us.',
  },
  emailLabel: {
    id: `${scope}.emailLabel`,
    defaultMessage: 'Your Email',
  },
  firstNameLabel: {
    id: `${scope}.firstNameLabel`,
    defaultMessage: 'First Name',
  },
  general: {
    id: `${scope}.general`,
    defaultMessage: 'General',
  },
  lastNameLabel: {
    id: `${scope}.lastNameLabel`,
    defaultMessage: 'Last Name',
  },
  langSettings: {
    id: `${scope}.langSettings`,
    defaultMessage: 'Your Language Settings',
  },
  minCharacters: {
    id: `${scope}.minCharacters`,
    defaultMessage: 'Field must be at least 6 characters',
  },
  invalidPassword: {
    id: `${scope}.invalidPassword`,
    defaultMessage: 'Invalid Password, Please Try Again.',
  },
  newPassword: {
    id: `${scope}.newPassword`,
    defaultMessage: 'New Password',
  },
  password: {
    id: `${scope}.password`,
    defaultMessage: 'Password',
  },
  pwdSettingsHeader: {
    id: `${scope}.pwdSettingsHeader`,
    defaultMessage: 'Change Your Password',
  },
  passwordReq: {
    id: `${scope}.passwordReq`,
    defaultMessage:
      'Please make sure your password is strong and 6 characters minimum.',
  },
  passwordChange: {
    id: `${scope}.passwordChange`,
    defaultMessage: 'To change your password please click the link below',
  },
  passwordUpdated: {
    id: `${scope}.passwordUpdated`,
    defaultMessage: 'Password Updated.',
  },
  passwordMatch: {
    id: `${scope}.passwordMatch`,
    defaultMessage: 'Passwords must match.',
  },
  passwordLink: {
    id: `${scope}.passwordLink`,
    defaultMessage: 'Change Your Password',
  },
  organization: {
    id: `${scope}.organization`,
    defaultMessage: 'Organization',
  },
});
