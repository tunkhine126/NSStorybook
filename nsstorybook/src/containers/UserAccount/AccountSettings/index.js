/**
 *
 * AccountSettings
 *
 */

import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { Form } from 'informed';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';

import Grid from '@material-ui/core/Grid';

import { changeLocale } from 'containers/LanguageProvider/actions';
import { makeSelectLocale } from 'containers/LanguageProvider/selectors';
import Loading from 'components/shared/Loading';
import ViewHeader from 'components/shared/ViewHeader';
import Content from 'components/shared/Content';
import ActionPanel, { loadingAction } from 'components/shared/ActionPanel';
import Notification from 'components/shared/Notification';

import { UPDATE_PROFILE } from 'graphql/mutations/user-mutation';
import { GET_USER } from 'graphql/queries/user-queries';
import { GET_USER_ORGANIZATIONS } from 'graphql/queries/organization-queries';
import { generalErrorHandler } from 'utils/error-handler';
import globalMessages from 'messages';
import messages from '../messages';

import General from './general';
import LangSettings from './languageSettings';
import Password from './password';
import { commonStyles, InputWrapper } from '../styles';
import { accSettingsStyles } from './styles';

export function AccountSettings({ intl, onLocaleToggle, locale, ...rest }) {
  const formApiRef = useRef();
  const classes = commonStyles();
  const accSetClasses = accSettingsStyles();

  const [messageState, setmessageState] = useState({
    message: '',
    open: false,
    type: 'success',
  });

  const [langState, setLangState] = useState({ item: '' });
  const [loadingState, setloadingState] = useState(false);
  const [survlangState, setSurvLangState] = useState({
    item: 'us_en',
  });

  const {
    loading: queryLoading,
    error: queryError,
    data: userData,
    client,
  } = useQuery(GET_USER);

  const { loading: orgsLoading, error: orgsError, data: orgsData } = useQuery(
    GET_USER_ORGANIZATIONS
  );

  const [orgsState, setOrgsState] = useState([]);

  const [updateProfile, { error: mutationError }] = useMutation(UPDATE_PROFILE);

  const messageHandler = (event, reason) => {
    if (reason === 'clickaway') return;
    setmessageState({ open: false });
  };

  function handleLangChange(e) {
    onLocaleToggle(e.target.value);
    setLangState({
      item: e.target.value,
    });
  }
  function handleSurvLangChange(e) {
    setSurvLangState({
      item: e.target.value,
    });
  }
  const sections = [
    {
      Component: General,
      formatedMsg: 'general',
      containerClass: classes.generalContainer,
      props: {
        orgsState,
      },
    },
    {
      Component: LangSettings,
      formatedMsg: 'langSettings',
      containerClass: accSetClasses.langSetContainer,
      props: {
        survlangState,
        langState,
        handleSurvLangChange,
        handleLangChange,
        setSurvLangState,
      },
    },
    {
      Component: Password,
      formatedMsg: 'password',
      containerClass: accSetClasses.passwordContainer,
    },
  ];
  const content = data =>
    sections.map(({ formatedMsg, Component, containerClass, props }, idx) => (
      <Content
        key={idx}
        subHeader={intl.formatMessage(messages[formatedMsg])}
        body={<Component data={data} state={props} />}
        containerClass={containerClass}
        headerClass={classes.sectionHeader}
      />
    ));

  const handleSave = async (data, callBack) => {
    data.defaultContextPermalink = survlangState.item;
    data.language = langState.item;
    delete data.organization;

    try {
      setloadingState(true);

      const response = await callBack({
        variables: { ...data },
      });
      if (response.data) {
        const {
          firstName,
          lastName,
          defaultContext: { permalink },
          language,
        } = {
          ...response.data.updateProfile.viewer,
        };

        // Upon successful save, update apollo cache
        client.writeData({
          data: {
            viewer: {
              firstName,
              lastName,
              defaultContextPermalink: permalink,
              language,
              __typename: 'User',
            },
          },
        });

        setmessageState(state => ({
          ...state,
          open: true,
          message: intl.formatMessage(messages.accountUpdated),
        }));
        const timeOut = loadingAction(setmessageState, setloadingState);

        // Needed to prevent memory leak when component unmounts
        clearTimeout(timeOut);
      }
    } catch (error) {
      const { graphQLErrors, networkError } = error;

      if (graphQLErrors.length > 0) {
        const [{ message }] = graphQLErrors;
        setmessageState({
          message,
          open: true,
          type: 'error',
        });
      }

      if (networkError) {
        const { message } = networkError;
        const failedConnection = message.includes('Failed to fetch');

        if (failedConnection) {
          setmessageState({
            message: intl.formatMessage(globalMessages.networkError),
            open: true,
            type: 'error',
          });
        }
      }
    }
  };

  const handleFormReset = () => {
    formApiRef.current.reset();
  };

  useEffect(() => {
    setLangState({
      item: locale,
    });

    if (!orgsLoading && orgsData) {
      setOrgsState(orgsData.organizations);
    }
  }, [locale, orgsData, orgsLoading]);

  if (queryLoading || orgsLoading || orgsState.length === 0) return <Loading />;

  if (queryError || orgsError || mutationError) {
    return generalErrorHandler(queryError || orgsError || mutationError, {
      ...rest,
    });
  }

  return (
    <>
      <ViewHeader
        component="Account Settings"
        body={intl.formatMessage(messages.accSettingsHeader)}
      />
      <div className={classes.root}>
        <Grid container>
          <Notification
            status={messageState.open}
            message={messageState.message}
            classes={classes}
            messageHandler={messageHandler}
            duration={3000}
            position={{ vertical: 'top', horizontal: 'right' }}
            type={messageState.type}
          />
          <Grid item xs={9} xl={10} className={classes.settingsContainer}>
            <InputWrapper>
              <Form
                getApi={formApi => {
                  formApiRef.current = formApi;
                }}
                id="update-account"
                onSubmit={formData => handleSave(formData, updateProfile)}
              >
                {content(userData)}
              </Form>
            </InputWrapper>
          </Grid>
          <ActionPanel
            action="save"
            type="submit"
            formId="update-account"
            resetForm={handleFormReset}
            loading={loadingState}
            disabled={loadingState}
          />
        </Grid>
      </div>
    </>
  );
}

AccountSettings.propTypes = {
  intl: intlShape.isRequired,
  onLocaleToggle: PropTypes.func,
  locale: PropTypes.string,
};

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale,
  })
);

const intlAccountSettings = injectIntl(AccountSettings);

export function mapDispatchToProps(dispatch) {
  return {
    onLocaleToggle: value => dispatch(changeLocale(value)),
    dispatch,
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(intlAccountSettings);
