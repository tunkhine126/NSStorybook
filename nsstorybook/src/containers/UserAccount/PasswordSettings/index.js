/**
 *
 * PasswordSettings
 *
 */

import React, { useRef, useState } from 'react';
import { injectIntl, intlShape } from 'react-intl';
import { Form } from 'informed';
import { useMutation } from '@apollo/react-hooks';

import Grid from '@material-ui/core/Grid';

import Content from 'components/shared/Content';
import ActionPanel, { loadingAction } from 'components/shared/ActionPanel';
import ViewHeader from 'components/shared/ViewHeader';
import Notification from 'components/shared/Notification';
import Loading from 'components/shared/Loading';

import { UPDATE_PASSWORD } from 'graphql/mutations/user-mutation';
import { generalErrorHandler } from 'utils/error-handler';
import globalMessages from 'messages';
import NewPassword from './newPassword';
import ConfirmChanges from './confirmChanges';
import messages from '../messages';

import { commonStyles, InputWrapper } from '../styles';
import { pwdSettingsStyles } from './styles';

export function PasswordSettings({ intl, ...rest }) {
  const formApiRef = useRef();
  const classes = commonStyles();
  const pwdSettingsClasses = pwdSettingsStyles();

  const [loadingState, setloadingState] = useState(false);
  const [messageState, setmessageState] = useState({
    message: '',
    open: false,
    type: 'success',
  });

  const [updatePassword, { loading, error }] = useMutation(UPDATE_PASSWORD);

  const messageHandler = (event, reason) => {
    if (reason === 'clickaway') return;
    setmessageState({ open: false });
  };

  const handleFormReset = () => {
    formApiRef.current.reset();
    formApiRef.current.setError('password', '');
  };

  const handleSubmit = async (data, callBack) => {
    try {
      setloadingState(true);

      const response = await callBack({
        variables: { ...data },
      });
      if (response.data) {
        setmessageState({
          message: intl.formatMessage(messages.passwordUpdated),
          open: true,
          type: 'success',
        });
        const timeOut = loadingAction(setmessageState, setloadingState);

        // Needed to prevent memory leak when component unmounts
        clearTimeout(timeOut);
        handleFormReset();
      }
    } catch (error) {
      const { graphQLErrors, networkError } = error;
      setloadingState(false);

      if (graphQLErrors.length > 0) {
        const [{ message }] = graphQLErrors;
        const invalidPassword = message.includes('Not authorized');

        setmessageState({
          message: invalidPassword
            ? intl.formatMessage(messages.invalidPassword)
            : message,
          open: true,
          type: 'error',
        });
      }

      if (networkError) {
        const { message } = networkError;
        const failedConnection = message.includes('Failed to fetch');
        if (failedConnection) {
          setmessageState({
            message: intl.formatMessage(globalMessages.networkError),
            open: true,
            type: 'error',
          });
        }
      }
    }
  };

  if (loading) return <Loading />;
  if (error) {
    return generalErrorHandler(error, { ...rest });
  }

  return (
    <>
      <ViewHeader
        component="Password Settings"
        body={intl.formatMessage(messages.pwdSettingsHeader)}
      />

      <div className={classes.root}>
        <Grid container>
          <Notification
            status={messageState.open}
            message={messageState.message}
            classes={classes}
            messageHandler={messageHandler}
            duration={3000}
            position={{ vertical: 'top', horizontal: 'right' }}
            type={messageState.type}
          />
          <Grid item xs={9} xl={10} className={classes.settingsContainer}>
            <InputWrapper>
              <Form
                getApi={formApi => {
                  formApiRef.current = formApi;
                }}
                id="change-password"
                onSubmit={formData => handleSubmit(formData, updatePassword)}
              >
                <Content
                  subHeader={intl.formatMessage(messages.newPassword)}
                  body={<NewPassword />}
                  containerClass={pwdSettingsClasses.newPassword}
                  headerClass={classes.sectionHeader}
                />
                <Content
                  subHeader={intl.formatMessage(messages.confirmChanges)}
                  body={<ConfirmChanges />}
                  containerClass={pwdSettingsClasses.confirmContainer}
                  headerClass={classes.sectionHeader}
                />
              </Form>
            </InputWrapper>
          </Grid>
          <ActionPanel
            action="update"
            type="submit"
            formId="change-password"
            resetForm={handleFormReset}
            loading={loadingState}
            disabled={loadingState}
          />
        </Grid>
      </div>
    </>
  );
}

PasswordSettings.propTypes = {
  intl: intlShape.isRequired,
};

const intlPasswordSettings = injectIntl(PasswordSettings);

export default intlPasswordSettings;
