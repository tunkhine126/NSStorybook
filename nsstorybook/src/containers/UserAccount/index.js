/**
 *
 * UserAccount
 *
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';

import messages from './messages';

export function UserAccount() {
  return (
    <div>
      <Helmet>
        <title>New Story - Account</title>
        <meta name="description" content="User Account" />
      </Helmet>
      <FormattedMessage {...messages.header} />
    </div>
  );
}

export default UserAccount;
