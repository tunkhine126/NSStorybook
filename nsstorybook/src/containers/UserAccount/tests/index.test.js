/**
 *
 * Tests for UserAccount
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import TestingComponent from 'components/shared/Testing';

import UserAccount from '../Loadable';

describe('<UserAccount />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const dispatch = jest.fn();
    render(
      <TestingComponent>
        <UserAccount dispatch={dispatch} />
      </TestingComponent>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const dispatch = jest.fn();
    const {
      container: { firstChild },
    } = render(
      <TestingComponent>
        <UserAccount dispatch={dispatch} />
      </TestingComponent>
    );
    expect(firstChild).toMatchSnapshot();
  });
});
