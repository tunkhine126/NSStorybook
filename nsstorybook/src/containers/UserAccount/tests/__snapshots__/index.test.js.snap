// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAccount /> Should render and match the snapshot 1`] = `
<div>
  <span>
    This is the UserAccount container!
  </span>
</div>
`;
